src/main/java/com/jfrog/ide/idea/scan/ScannerFactory.java

```diff
@@ -0,0 +1,192 @@
+package com.jfrog.ide.idea.scan;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Maps;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.projectRoots.Sdk;
+import com.jetbrains.python.sdk.PythonSdkUtil;
+import com.jfrog.ide.common.scan.ScanLogic;
+import com.jfrog.ide.common.utils.PackageFileFinder;
+import com.jfrog.ide.idea.configuration.GlobalSettings;
+import com.jfrog.ide.idea.log.Logger;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+import static com.jfrog.ide.idea.scan.ScanUtils.createScanPaths;
+import static com.jfrog.ide.idea.utils.Utils.getProjectBasePath;
+
+/**
+ * Created by yahavi
+ */
+public class ScannerFactory {
+    private final Project project;
+
+    public ScannerFactory(Project project) {
+        this.project = project;
+    }
+
+    public static int getModuleIdentifier(String name, String path) {
+        return Objects.hashCode(name, path);
+    }
+
+    /**
+     * Scan projects, create new Scanners and delete unnecessary ones.
+     * Existing Scanners from previous scans, are not overridden.
+     */
+    public Map<Integer, ScannerBase> refreshScanners(Map<Integer, ScannerBase> oldScanners, ScanLogic scanLogic,
+                                                     @Nullable ExecutorService executor) throws IOException {
+        Map<Integer, ScannerBase> scanners = Maps.newHashMap();
+        refreshMavenScanner(scanners, oldScanners, executor, scanLogic);
+        refreshPypiScanners(scanners, oldScanners, executor, scanLogic);
+        Set<Path> scanPaths = createScanPaths(oldScanners, project);
+        refreshGenericScanners(scanners, oldScanners, scanPaths, executor, scanLogic);
+        return scanners;
+    }
+
+    /**
+     * Create npm, Gradle, Go and Yarn Scanners.
+     *
+     * @param newScanners the new Scanners map to add the Scanners into
+     * @param oldScanners the Scanners map including the Scanner of the current project, or an empty map for a fresh start
+     * @param scanPaths   potential paths for scanning for package descriptor files
+     * @param executor    the thread pool
+     * @throws IOException in case of any I/O error during the search for the actual package descriptor files.
+     */
+    private void refreshGenericScanners(Map<Integer, ScannerBase> newScanners, Map<Integer, ScannerBase> oldScanners,
```

Comment:
is there any advantage for the "output" param` newScanners`
(instead of returning it as a return value)?

Reply:
It's not exactly an output param. `newScanners` is created outside of this method, and even might contain Scanners in it.